dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/ApplicationProcessor.java

```diff
@@ -0,0 +1,591 @@
+import com.ibm.ws.container.service.app.deploy.EARApplicationInfo;
+import com.ibm.ws.container.service.app.deploy.WebModuleInfo;
+import com.ibm.ws.ffdc.annotation.FFDCIgnore;
+import com.ibm.wsspi.adaptable.module.Container;
+import com.ibm.wsspi.adaptable.module.Entry;
+import com.ibm.wsspi.adaptable.module.NonPersistentCache;
+import com.ibm.wsspi.adaptable.module.UnableToAdaptException;
+import com.ibm.wsspi.http.VirtualHost;
+
+import io.openliberty.microprofile.openapi20.utils.CloudUtils;
+import io.openliberty.microprofile.openapi20.utils.Constants;
+import io.openliberty.microprofile.openapi20.utils.IndexUtils;
+import io.openliberty.microprofile.openapi20.utils.LoggingUtils;
+import io.openliberty.microprofile.openapi20.utils.MessageConstants;
+import io.openliberty.microprofile.openapi20.utils.ModuleUtils;
+import io.openliberty.microprofile.openapi20.utils.OpenAPIUtils;
+import io.openliberty.microprofile.openapi20.utils.ProxySupportUtil;
+import io.openliberty.microprofile.openapi20.utils.ServerInfo;
+import io.openliberty.microprofile.openapi20.utils.ValidationMessageConstants;
+import io.openliberty.microprofile.openapi20.validation.OASValidationResult;
+import io.openliberty.microprofile.openapi20.validation.OASValidationResult.ValidationEvent.Severity;
+import io.openliberty.microprofile.openapi20.validation.OASValidator;
+import io.openliberty.microprofile.openapi20.validation.ValidatorUtils;
+import io.smallrye.openapi.api.OpenApiConfig;
+import io.smallrye.openapi.api.OpenApiDocument;
+import io.smallrye.openapi.api.models.info.InfoImpl;
+import io.smallrye.openapi.runtime.OpenApiProcessor;
+import io.smallrye.openapi.runtime.OpenApiStaticFile;
+import io.smallrye.openapi.runtime.io.Format;
+import io.smallrye.openapi.runtime.io.OpenApiSerializer;
+
+/**
+ * The ApplicationProcessor class processes applications that are deployed to the OpenLiberty instance in order to
+ * generate OpenAPI documents. However, the MP OpenAPI functionality in OpenLiberty only supports generating an OpenAPI
+ * document for a single application at a time so, if multiple applications are deployed to the OpenLiberty instance,
+ * an OpenAPI document will only be generated for the first application that is processed. Also, if an enterprise
+ * application (EAR/EBA) is deployed that contains multiple web modules, an OpenAPI document will only be generated for
+ * the first Web Module that generates an OpenAPI document. 
+ */
+@Component(service = { ApplicationProcessor.class }, configurationPolicy = ConfigurationPolicy.IGNORE, immediate = true, property = { "service.vendor=IBM" })
+public class ApplicationProcessor {
+
+    private static final TraceComponent tc = Tr.register(ApplicationProcessor.class);
+
+    private static ApplicationProcessor instance = null;
+    private OpenAPI document = null;
+    private static ApplicationInfo currentApp = null;
+    private static Map<String, ApplicationInfo> applications = new HashMap<>();
```

Comment:
Unsure why these two fields are static.

There are multiple times where access to them is synchronized on `document` which is non-static.

I guess it shouldn't matter since this should be a singleton class anyway?

Reply:
Agreed... I am not completely happy with the implementation of the `ApplicationProcessor`. I would like to pull it apart so that it is solely focused on parsing an application and generating an OpenAPI document, if possible.  I would then store the state elsewhere.  This would mean that the `ApplicationProcessor` would no longer need to be an OSGi component.  However, it is not clear to me how much work this will entail and I want to get an initial implementation of the feature into GitHub... not least because the performance team want to compare it against previous versions.